File: src/main/java/hiiragi283/material/api/shape/HiiragiShape.java
package hiiragi283.material.api.shape;

import com.github.bsideup.jabel.Desugar;
import com.google.common.base.CaseFormat;
import hiiragi283.material.api.event.HiiragiRegistryEvent;
import hiiragi283.material.api.material.HiiragiMaterial;
import hiiragi283.material.api.registry.HiiragiRegistry;
import net.minecraft.client.resources.I18n;
import net.minecraftforge.common.MinecraftForge;
import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

@Desugar
public record HiiragiShape(String name, int scale) {

    @NotNull
    public String getOreDict(HiiragiMaterial material) {
        return getOreDictPrefix() + material.getOreDictName();
    }

    public List<String> getOreDicts(HiiragiMaterial material) {
        List<String> list = new ArrayList<>();
        list.add(getOreDict(material));
        if (!material.oreDictAlt.isEmpty()) {
            for (String oreDict : material.oreDictAlt) {
                list.add(getOreDictPrefix() + CaseFormat.LOWER_UNDERSCORE.to(CaseFormat.UPPER_CAMEL, oreDict));
            }
        }
        return list;
    }

    @NotNull
    public String getOreDictPrefix() {
        return CaseFormat.LOWER_UNDERSCORE.to(CaseFormat.LOWER_CAMEL, name);
    }

    @NotNull
    public String getTranslatedName(HiiragiMaterial material) {
        return I18n.format("hiiragi_shape." + name, material.getTranslatedName());
    }

    public boolean hasScale() {
        return scale > 0;
    }

    public boolean isValid(HiiragiMaterial material) {
        return material.shapeType.shapes().contains(this);
    }

    //    General    //

    @Override
    public boolean equals(Object obj) {
        if (obj == null) return false;
        if (!(obj instanceof HiiragiShape)) return false;
        return Objects.equals(((HiiragiShape) obj).name, this.name);
    }

    @Override
    public String toString() {
        return "Shape:" + name;
    }

    //    Registry    //

    public static HiiragiRegistry<String, HiiragiShape> REGISTRY = new HiiragiRegistry<>("Shape");

    public static void registerShapes() {
        var event = new HiiragiRegistryEvent.Shape(HiiragiShape.REGISTRY);
        MinecraftForge.EVENT_BUS.post(event);
        HiiragiShape.REGISTRY.setUnmodifiable();
    }

}

File: src/main/kotlin/hiiragi283/material/api/shape/HiiragiShape.kt
package hiiragi283.material.api.shape

import com.google.gson.JsonElement
import com.google.gson.JsonObject
import hiiragi283.material.api.material.HiiragiMaterial
import hiiragi283.material.api.part.HiiragiPart
import hiiragi283.material.init.HiiragiRegistries
import hiiragi283.material.init.HiiragiShapeTypes
import hiiragi283.material.util.HiiragiJsonSerializable
import hiiragi283.material.util.itemStack
import net.minecraft.client.resources.I18n
import net.minecraft.item.ItemStack
import net.minecraftforge.oredict.OreDictionary
import rechellatek.snakeToLowerCamelCase

/**
 * An object which represents the shape of [net.minecraft.item.Item]
 *
 * Should be registered in [net.minecraftforge.event.RegistryEvent.Register]<[HiiragiShape]>
 */

data class HiiragiShape(val name: String, val scale: Int) : HiiragiJsonSerializable {

    fun getItemStack(material: HiiragiMaterial, count: Int = 1): ItemStack =
        HiiragiRegistries.MATERIAL_ITEM.getValue(this)?.item()?.itemStack(material, count) ?: ItemStack.EMPTY

    fun getItemStack(part: HiiragiPart): ItemStack {
        val scale: Int = part.shape.scale
        return if (scale >= 144) getItemStack(part.material, scale / 144) else ItemStack.EMPTY
    }

    fun getItemStackWild(count: Int = 1): ItemStack = getItemStack(HiiragiMaterial.WILDCARD, count)

    fun getItemStacks(count: Int = 1): List<ItemStack> = HiiragiRegistries.MATERIAL_INDEX.getValues()
        .map(::getPart)
        .flatMap { it.getItemStacks(count) }

    fun getOreDict(material: HiiragiMaterial): String = name.snakeToLowerCamelCase() + material.getOreDictName()

    fun getOreDicts(material: HiiragiMaterial): List<String> {
        val list: MutableList<String> = mutableListOf()
        list.add(getOreDict(material))
        if (material.hasOreDictAlt()) {
            material.getOreDictNameAlt().forEach { oreDict ->
                list.add(StringBuilder().also {
                    it.append(name.snakeToLowerCamelCase())
                    it.append(oreDict)
                }.toString())
            }
        }
        return list.filter(String::isNotEmpty)
    }

    fun getPart(material: HiiragiMaterial): HiiragiPart = HiiragiPart(this, material)

    fun getTranslatedName(material: HiiragiMaterial): String =
        I18n.format("hiiragi_shape.$name", material.getTranslatedName())

    fun hasScale(): Boolean = scale > 0

    fun isValid(material: HiiragiMaterial): Boolean =
        material.shapeType == HiiragiShapeTypes.WILDCARD || this in material.shapeType.shapes

    fun hasValidItem(material: HiiragiMaterial): Boolean = getOreDicts(material).any(OreDictionary::doesOreNameExist)

    override fun toString(): String = "Shape:$name"

    //    Registration    //

    fun register() {
        HiiragiRegistries.SHAPE.register(name, this)
    }

    //    HiiragiJsonSerializable    //

    override fun getJsonElement(): JsonElement {
        val root = JsonObject()
        root.addProperty("name", name)
        root.addProperty("scale", scale)
        return root
    }

}